libass/ass_fontselect.c

```diff
@@ -166,6 +170,111 @@ static ASS_FontProviderFuncs ft_funcs = {
+/*
+ * Output parameters are only valid if the call was sucessful, ignore them otherwise.
+ * \param path          input; path of the dir to open, may be UTF-8 or the codepage used by fopen
+ * \param file_context  output
+ * \param find_handle   output; needs to be closed
+ * \param path_utf8     output; utf-8 encoded version of path; needs to be freed
+ * \return true if succesful, false otherwise
+ */
+static wchar_t *win_open_dir(const char *path, WIN32_FIND_DATAW *file_context, HANDLE *find_handle, size_t *out_length)
+{
+    UINT cur_cp = CP_UTF8;
+    size_t tmp;
+
+    if (!path || !(tmp = strlen(path)) || tmp > SIZE_MAX - 3)
+        return NULL;
+
+    char* valid_search_dir = malloc(tmp + 3);
+    if (!valid_search_dir)
+        return NULL;
+    strcpy(valid_search_dir, path);
+    switch (valid_search_dir[tmp - 1]) {
+    case '\\':
+    case '/':
+        break;
+    default:
+        valid_search_dir[tmp++] = '\\';
+    }
+    valid_search_dir[tmp++] = '*';
+    valid_search_dir[tmp] = '\0';
+
+    size_t wlen;
+    wchar_t *wpath = ass_convert_to_utf16le(valid_search_dir, cur_cp, &wlen);
+    if (!wpath || (*find_handle = FindFirstFileExW(wpath, 0, file_context, 0, NULL, 0)) == INVALID_HANDLE_VALUE) {
+        free(wpath);
+        cur_cp = CP_ACP;
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
+        if (!AreFileApisANSI()) cur_cp = CP_OEMCP;
+#endif
+        wpath = ass_convert_to_utf16le(valid_search_dir, cur_cp, &wlen);
+        if (wpath) {
+            if ((*find_handle = FindFirstFileExW(wpath, 0, file_context, 0, NULL, 0)) == INVALID_HANDLE_VALUE) {
+                free(wpath);
+                wpath = NULL;
+                goto clean;
+            }
+        }
+        else
+            goto clean;
+    }
+
+    // path_utf8 ends with "\*" or "/*"
+    wlen -= 2;
+    wpath[wlen] = '\0';
+    if (out_length)
+        *out_length = wlen;
+
+clean:
+    free(valid_search_dir);
+    return wpath;
+}
+
+static void load_fonts_from_dir(ASS_Library *library, const char *dir)
+{
+    WIN32_FIND_DATAW find_file_context;
+    HANDLE find_file_handle;
+    wchar_t *wdir = NULL;
+    size_t wdirlen;
+
+    if (!(wdir = win_open_dir(dir, &find_file_context, &find_file_handle, &wdirlen)))
+        return;
+
+    char *dir_utf8 = ass_convert_from_utf16le(wdir, CP_UTF8, NULL);
+
+    size_t namemax = 0;
+    wchar_t *wnamebuf = NULL;
+    do {
+        if (find_file_context.cFileName[0] == L'.')
+            continue;
+        size_t namelen = wdirlen + wcslen(find_file_context.cFileName) + 2u;
+        if (namelen < 2 || namelen - 2 < wdirlen)
+            continue;
+        if (namelen > namemax) {
+            size_t newlen = FFMAX(2048, namelen + FFMIN(256, SIZE_MAX - namelen));
+            if (ASS_REALLOC_ARRAY(wnamebuf, newlen))
+                namemax = newlen;
+            else
+                continue;
+        }
+        swprintf(wnamebuf, namemax, L"%s\\%s", wdir, find_file_context.cFileName);
```

Comment:
The format specifiers are wrong, they need to be `%ls` since the arguments are wide/long strings, not regular `char*` ones. *(And if they were `char*`, you'd need to check for `swnprintf`-failures during conversion.)*

Reply:
Microsoft’s `swprintf` uses `%s` for wide strings. But `%ls` should also work.